pkg/migrations/backfill.go

```diff
@@ -107,23 +111,95 @@ func (b *batcher) updateBatch(ctx context.Context, conn db.DB) error {
 	})
 }
 
+type batchStatementBuilder struct {
+	tableName       string
+	identityColumns []string
+	batchSize       int
+}
+
+func newBatchStatementBuilder(tableName string, identityColumnNames []string, batchSize int) *batchStatementBuilder {
+	quotedCols := make([]string, len(identityColumnNames))
+	for i, col := range identityColumnNames {
+		quotedCols[i] = pq.QuoteIdentifier(col)
+	}
+	return &batchStatementBuilder{
+		tableName:       pq.QuoteIdentifier(tableName),
+		identityColumns: quotedCols,
+		batchSize:       batchSize,
+	}
+}
+
 // buildQuery builds the query used to update the next batch of rows.
-func (b *batcher) buildQuery() string {
+func (sb *batchStatementBuilder) buildQuery(lastValues any) string {
+	return fmt.Sprintf("WITH batch AS (%[1]s), update AS (%[2]s) %[3]s",
+		sb.buildBatchSubQuery(lastValues),
+		sb.buildUpdateBatchSubQuery(),
+		sb.buildLastValueQuery())
+}
+
+// fetch the next batch of PK of rows to update
+func (sb *batchStatementBuilder) buildBatchSubQuery(lastValues any) string {
 	whereClause := ""
-	if b.lastValue != nil {
-		whereClause = fmt.Sprintf("WHERE %s > %v", pq.QuoteIdentifier(b.identityColumn.Name), pq.QuoteLiteral(*b.lastValue))
+	if lastValues != nil {
+		conditions := make([]string, len(sb.identityColumns))
+		switch lastVals := lastValues.(type) {
+		case []int64:
+			for i, col := range sb.identityColumns {
+				conditions[i] = fmt.Sprintf("%s > %d ", col, lastVals[i])
+			}
+		case []string:
+			for i, col := range sb.identityColumns {
+				conditions[i] = fmt.Sprintf("%s > %s ", col, pq.QuoteLiteral(lastVals[i]))
+			}
+		case []any:
+			for i, col := range sb.identityColumns {
+				if v, ok := lastVals[i].(int); ok {
+					conditions[i] = fmt.Sprintf("%s > %d ", col, v)
+				} else if v, ok := lastVals[i].(string); ok {
+					conditions[i] = fmt.Sprintf("%s > %s ", col, pq.QuoteLiteral(v))
+				} else {
+					panic("unsupported type")
+				}
+			}
+		case int64:
+			conditions[0] = fmt.Sprintf("%s > %d ", sb.identityColumns[0], lastVals)
+		case string:
+			conditions[0] = fmt.Sprintf("%s > %s ", sb.identityColumns[0], pq.QuoteLiteral(lastVals))
+		default:
+			panic("unsupported type")
+		}
+		whereClause = "WHERE " + strings.Join(conditions, "AND ")
 	}
 
-	return fmt.Sprintf(`
-    WITH batch AS (
-      SELECT %[1]s FROM %[2]s %[4]s ORDER BY %[1]s LIMIT %[3]d FOR NO KEY UPDATE
-    ), update AS (
-      UPDATE %[2]s SET %[1]s=%[2]s.%[1]s FROM batch WHERE %[2]s.%[1]s = batch.%[1]s RETURNING %[2]s.%[1]s
-    )
-    SELECT LAST_VALUE(%[1]s) OVER() FROM update
-    `,
-		pq.QuoteIdentifier(b.identityColumn.Name),
-		pq.QuoteIdentifier(b.table.Name),
-		b.batchSize,
-		whereClause)
+	return fmt.Sprintf("SELECT %[1]s FROM %[2]s %[3]sORDER BY %[1]s LIMIT %[4]d FOR NO KEY UPDATE",
```

Comment:
```suggestion
	return fmt.Sprintf("SELECT %[1]s FROM %[2]s %[3]s ORDER BY %[1]s LIMIT %[4]d FOR NO KEY UPDATE",
```

Reply:
For the record, this change adds an extra whitespace before `ORDER BY` if the where clause is missing.

Reply:
For the record this change adds an extra whitespace when `where` cluase is missing.